psdisk/psdisk.c

```diff
@@ -12,156 +12,570 @@
+		printf("The file contains incorrect partition table.\n");
+		return -1;
+	}
+
+	for (i = 0; i < pCnt; ++i) {
+		if ((pNode = (struct plist_node_t *)calloc(1, sizeof(struct plist_node_t))) == NULL) {
+			printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+			free(pHeaders);
+			return -1;
+		}
+		pNode->pHeader = pHeaders[i];
+		LIST_INSERT_HEAD(&psdisk_common.list, pNode, ptrs);
+		++psdisk_common.pCnt;
+	}
+
+	free(pHeaders);
+
+	return 0;
 }
 
 
-static int psdisk_createpTable(void)
+static int psdisk_verifyPartsTable(void)
 {
-	psdisk_common.mem.write = psdisk_writeData;
+	int i;
+	uint32_t pCnt;
+	struct plist_node_t *pNode;
+	ptable_partition_t *pHeaders;
+
+	/* Verify whether data was saved correctly */
+	if ((pHeaders = ptable_readPartitions(&pCnt, &psdisk_common.mem)) == NULL) {
+		printf("The file contains incorrect partition table.\n");
+		return -1;
+	}
 
-	/* TODO: call ptable_writePartitions function */
+	/* Verify values from file after writing */
+	i = 0;
+	LIST_FOREACH(pNode, &psdisk_common.list, ptrs) {
+		if (pNode->status == part_save) {
+			if (i >= psdisk_common.pCnt || memcmp(&pNode->pHeader, &pHeaders[i], sizeof(ptable_partition_t)) != 0) {
+				free(pHeaders);
+				return -1;
+			}
+			++i;
+		}
+	}
 
+	free(pHeaders);
 
 	return 0;
 }
 
 
-static int psdisk_parsePartitionOpt(int optind, char *argv[])
+static int psdisk_createPartsTable(void)
 {
 	int i;
-	char *name;
-	ptable_partition_t pHeader;
+	struct plist_node_t *node;
+	ptable_partition_t *pHeaders;
+
+	if ((pHeaders = (ptable_partition_t *)calloc(psdisk_common.pCnt, sizeof(ptable_partition_t))) == NULL) {
+		printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	/* Copy partition table to temporary array */
+	i = 0;
+	LIST_FOREACH(node, &psdisk_common.list, ptrs) {
+		if (node->status == part_save)
+			memcpy(&pHeaders[i++], &node->pHeader, sizeof(ptable_partition_t));
+	}
+
+	if (ptable_writePartitions(pHeaders, psdisk_common.pCnt, &psdisk_common.mem) < 0) {
+		printf("Cannot write partition table to file %s.\n", psdisk_common.fileName);
+		free(pHeaders);
+		return -1;
+	}
+
+	free(pHeaders);
+
+	return 0;
+}
 
-	const int MAX_PARTITION_ATTR = 4;
 
-	for (i = 0; i < MAX_PARTITION_ATTR; ++i) {
-		if (argv[optind + i] == NULL) {
-			LOG_ERROR("Invalid number of arguments.");
+static int psdisk_createImg(void)
+{
+	if (psdisk_createPartsTable() < 0)
+		return -1;
+
+	if (psdisk_verifyPartsTable() < 0)
+		return -1;
+
+	psdisk_showPartsTable();
+
+	return 0;
+}
+
+
+static int psdisk_updatePartsList(ptable_partition_t *pHeaders, uint32_t pCnt)
+{
+	int i, action, namesEq;
+	struct plist_node_t *pNode, *newNode, *tempNode;
+
+	for (i = 0; i < pCnt; ++i) {
+		/* Default action: partition from file differs from partition defines by user. */
+		action = part_save;
+
+		LIST_FOREACH(pNode, &psdisk_common.list, ptrs) {
+			namesEq = (strcmp((const char *)&pNode->pHeader.name, (const char *)&pHeaders[i].name) == 0);
```

Comment:
No need for this variable, really. You can wrap whole body in if (strcmp(.... == 0)

Reply:
Also, are we sure we get null terminator? Maybe strncmp for extra safety?

Reply:
Declaring value from function to variable should be faster than calling the same function two times. It will be also copy-paste. `pNode->pHeader.name` and `pHeaders[i].name` are `unsigned char name[8]`. The size is pre defined so I think there is no need for additional checking.

Reply:
if (strcmp((const char *)&pNode->pHeader.name, (const char *)&pHeaders[i].name) == 0)
{
if (pNode->status == part_remove) {
...
}
else if (pNode->status == part_save) {
..
}
}

I was asking about null terminator, not size. I know size is const, so we can easly use strncmp if we aren't so sure about null terminator.